### PR TITLE
Change TaskSpec to allow multiple object IDs per argument.

### PR DESCRIPTION
**Pull request by pcmoritz**
This code was contributed by @imzhenyu

The purpose of this is to make it possible to implement a MapReduce framework on top of Ray. In this case, each reducer can get inputs from every mapper. This PR implements passing lists of object ids for every argument to support this use case.

In the future we plan to change this and have a single list arg_ids of object IDs (which would be the concatentation of all the lists from this PR and is the only information taken into account by the scheduler) and a data blob arg_data which will allow to store arbitrary structure, e.g. the nesting of the lists in this case.

The list `[[obj_id1, obj_id2, obj_id3], [obj_id4, obj_id5]]` would be represented like so: `arg_data = [[obj_id1, obj_id2, obj_id3], [obj_id4, obj_id5]]` and `arg_ids = [obj_id1, obj_id2, obj_id3, obj_id4, obj_id5]`

### CODE REVIEW

File: src/common/format/common.fbs [hunk 1]
```diff
@@ -17,8 +17,11 @@ enum ResourceIndex:int {
 }
 
 table Arg {
-  // Object ID for pass-by-reference arguments.
-  object_id: string;
+  // Object ID for pass-by-reference arguments. Normally there is only one
+  // object id in this list which represents the object that is being passed.
```
Comment (robertnishihara) [inline src/common/format/common.fbs]: id -> ID

File: src/common/task.h [hunk 2]
```diff
@@ -239,14 +239,24 @@ int64_t TaskSpec_num_returns(TaskSpec *spec);
 bool TaskSpec_arg_by_ref(TaskSpec *spec, int64_t arg_index);
 
 /**
+ * Get number of object ids in a given argument
```
Comment (robertnishihara) [inline src/common/task.h]: ids -> IDs

File: src/common/task.h [hunk 3]
```diff
@@ -239,14 +239,24 @@ int64_t TaskSpec_num_returns(TaskSpec *spec);
...
+ *
+ * @param spec The task_spec in question.
+ * @param arg_index The index of the argument in question.
+ * @return number of object ids in this argument
```
Comment (robertnishihara) [inline src/common/task.h]: ids -> IDs

File: src/common/task.h [hunk 4]
```diff
@@ -272,11 +282,14 @@ int64_t TaskSpec_arg_length(TaskSpec *spec, int64_t arg_index);
  * arguments in their order of appearance.
  *
  * @param spec The task_spec in question.
- * @param The object ID to set the argument to.
+ * @param object_ids The object IDs to set the argument to.
+ * @param num_object_ids number of ids in this param, usually 1.
```
Comment (robertnishihara) [inline src/common/task.h]: ids -> IDs

File: src/global_scheduler/global_scheduler_algorithm.cc [hunk 5]
```diff
@@ -94,43 +94,40 @@ int64_t locally_available_data_size(const GlobalSchedulerState *state,
...
-      /* This local scheduler does not have access to this object, so don't
-       * count this object. */
-      continue;
+    int count = TaskSpec_arg_id_count(task_spec, i);
+    for (int j = 0; j < count; ++j) {
+      ObjectID object_id = TaskSpec_arg_id(task_spec, i, j);
+
+      if (state->scheduler_object_info_table.count(object_id) == 0) {
+        /* If this global scheduler is not aware of this object ID, then ignore
+         * it.*/
```
Comment (robertnishihara) [inline src/global_scheduler/global_scheduler_algorithm.cc]: add a space before `*/`

File: src/common/task.h [hunk 6]
```diff
@@ -239,14 +239,24 @@ int64_t TaskSpec_num_returns(TaskSpec *spec);
...
+ */
+int TaskSpec_arg_id_count(TaskSpec *spec, int64_t arg_index);
+
+/**
  * Get a particular argument to this task. This assumes the argument is an
  * object ID.
  *
  * @param spec The task_spec in question.
  * @param arg_index The index of the argument in question.
+ * @param id_index The index of the object index in this arg.
```
Comment (robertnishihara) [inline src/common/task.h]: object index -> object ID?

**Comment by robertnishihara** [comment]
This is a workaround for languages (e.g., Java, which do not allow functions to have a variable number of arguments, right? For example, this would not be needed in order to implement MapReduce in Python, right?

**Comment by pcmoritz** [comment]
This is right, it is a workaround for Java in this case; I also added a paragraph above explaining how we should do it in the future.

**Comment by robertnishihara** [comment]
retest this please

**Comment by pcmoritz** [comment]
retest this please

**Review by robertnishihara — APPROVED** [review]

**Pull request merged**